criu/cr-restore.c

```diff
@@ -3128,6 +3128,24 @@ static void rst_reloc_creds(struct thread_restore_args *thread_args,
 	thread_args->creds_args = args;
 }
 
+static bool groups_are_okay(gid_t* groups, int n_groups) {
+	int n, len;
+	bool ret;
+	gid_t* gids;
+
+	n = getgroups(0, NULL);
+	if (n != n_groups)
+		return false;
+	if (n == 0)
+		return true;
+	len = n * sizeof(gid_t);
+	gids = (gid_t*)xmalloc(len);
+	n = getgroups(n, gids);
```

Comment:
`getgroups` may fail, it's return value should be verified.

Reply:
We **can** check if `getgroups` fails here, but the Linux man page for `getgroups` states that it will fail either if gids has an invalid address or if n is "less than the number of supplementary group IDs, but is not zero." 
Given this, I'll be sure to follow your other suggestions to ensure that gids is valid -- however, `getgroups` should not fail here after those changes.

Reply:
It may also fail with `-ENOMEM` or `-EPERM`:

```
ENOMEM Out of memory.
EPERM  The calling process has insufficient privilege (the caller  does not have the CAP_SETGID capability in the user namespace in which it resides).
EPERM (since Linux 3.19) The use of setgroups() is denied in this user namespace.  See the description of /proc/[pid]/setgroups in user_namespaces(7).
```

https://man7.org/linux/man-pages/man2/getgroups.2.html

Reply:
That man page contains the possible failures for both `getgroups` and `setgroups`. You are absolutely correct in that `setgroups` could fail for additional reasons, but I'd defer to the `getgroups` failures for this line of code.

Reply:
You are right, I've skipped over the line "setgroups() can additionally fail" :)

Still, checking a system call return value is a good practice even if you know for sure that there should be no error. 
Some code around this call may change and this assumption will not hold anymore, or the kernel implementation will change and it may return error in some case that was not anticipated.

It is better to have the check for the system call return value from the beginning to be on the safe side.